### PR TITLE
Add azurerm_storage_data_lake_gen2_path with support for folders and ACLs

### PR DESCRIPTION
**Pull request by stuartleeks**
~~NOTE that this PR currently has a commit to add in the vendored code for [this PR](https://github.com/tombuildsstuff/giovanni/pull/28) (this will be rebased out once the PR is merged)~~

This PR adds the start of the `azurerm_storage_data_lake_gen2_path` resource (#7118) with support for creating folders and ACLs as per [this comment](https://github.com/terraform-providers/terraform-provider-azurerm/issues/7118#issuecomment-651161762).

### CODE REVIEW

**Comment by stuartleeks** [comment]
Rebased and added support for setting folder ACLs (and updated the PR comment above)

Would welcome review of this PR to give time to make any changes so that it is ready for when the corresponding [giovanni PR](https://github.com/tombuildsstuff/giovanni/pull/28) is merged :-)

**Comment by stuartleeks** [comment]
Rebased now that [giovanni is updated to v0.11.0](https://github.com/terraform-providers/terraform-provider-azurerm/pull/7608)

**Comment by stuartleeks** [comment]
Rebased on latest master and fixed up CI errors

**Comment by favoretti** [comment]
@tombuildsstuff Can this be shipped with 2.19 as well please?

File: azurerm/internal/services/storage/tests/resource_arm_storage_data_lake_gen2_path_test.go
Comment (jackofallops) [inline azurerm/internal/services/storage/tests/resource_arm_storage_data_lake_gen2_path_test.go:187]: This will need unpacking for something useful, or there will just be a memory address and a collection of empty fields.

**Review by jackofallops — DISMISSED** [review]
Hi @stuartleeks 
Thanks for the PR, afraid I've only had chance to do a fairly quick review here, there are some comments below.  I ran the tests and, for me, they all fail. It looks like the `delete` func either doesn't work as expected, or needs to poll/wait for the operation to complete:
```
TestAccAzureRMStorageDataLakeGen2Path_basic: testing.go:745: Error destroying resource! WARNING: Dangling resources
may exist. The full state and error is shown below.
Error: Check failed: Path still exists: {Response:{Response:0xc001cf0360} ETag: LastModified:0001-01-01 00:00:00 +0000 UTC ResourceType: Owner: Group: ACL:}
```

Additionally, there appears to be a permissions issue in setting the ACLs via `SetAccessControl`:
```
TestAccAzureRMStorageDataLakeGen2Path_withSimpleACL: testing.go:684: Step 2 error: errors during apply:
Error: Error setting access control for Path "fstest" in File System "testpath" in Storage Account "acctestacc79k41": datalakestore.Client#SetAccessControl: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationPermissionMismatch" Message="This request is not authorized to perform this operation using this permission.\nRequestId:1698a1b8-201f-005e-0e29-5cf5a2000000\nTime:2020-07-17T11:02:39.7133752Z"
```

If you can address/investigate the above, I'll loop back asap to complete the review.

Thanks!

**Comment by stuartleeks** [comment]
@jackofallops  - thanks for your review. Weird about the tests as they were working locally when I pushed the changes. I'll have to have a dig in and see what's happening there.

I'm on vacation the next two weeks (and likely starting a new project when I get back) but will take a look at this when I get chance.

File: azurerm/internal/services/storage/resource_arm_storage_data_lake_gen2_path.go
Comment (stuartleeks) [reply azurerm/internal/services/storage/resource_arm_storage_data_lake_gen2_path.go:200]: ok

File: azurerm/internal/services/storage/resource_arm_storage_data_lake_gen2_path.go
Comment (stuartleeks) [reply azurerm/internal/services/storage/resource_arm_storage_data_lake_gen2_path.go:190]: ok, will remove the test

File: azurerm/internal/services/storage/resource_arm_storage_data_lake_gen2_path.go
Comment (stuartleeks) [reply azurerm/internal/services/storage/resource_arm_storage_data_lake_gen2_path.go:89]: Yes, as per the discussion [here](https://github.com/terraform-providers/terraform-provider-azurerm/issues/7118#issuecomment-651161762)

**Comment by jackofallops** [comment]
> @jackofallops - thanks for your review. Weird about the tests as they were working locally when I pushed the changes. I'll have to have a dig in and see what's happening there.
> 
> I'm on vacation the next two weeks (and likely starting a new project when I get back) but will take a look at this when I get chance.

Not a problem, it may be that there are permissions for your user/SP that are not implicit for a subscription owner / GA?  It wouldn't be the first time we've had to go dig for explicit permissions for the testing account.  If I get chance I'll look into it.

File: azurerm/internal/services/storage/resource_arm_storage_data_lake_gen2_path.go
Comment (stuartleeks) [reply azurerm/internal/services/storage/resource_arm_storage_data_lake_gen2_path.go:364]: ok - will inline it

**Comment by stuartleeks** [comment]
Can you share the test error that you saw? I'm wondering whether the test failed and didn't clean up, or something like that?

The test user needs to have the `Storage Blob Data Owner` permission, I think.

File: azurerm/internal/services/storage/tests/resource_arm_storage_data_lake_gen2_path_test.go
Comment (stuartleeks) [reply azurerm/internal/services/storage/tests/resource_arm_storage_data_lake_gen2_path_test.go:187]: I think the `path` would be more useful than the `props` there - will amend

**Comment by jackofallops** [comment]
> Can you share the test error that you saw? I'm wondering whether the test failed and didn't clean up, or something like that?
> 
> The test user needs to have the `Storage Blob Data Owner` permission, I think.

2 of the 5 test results (_basic, and _withSimpleACL) are included in the review note above, I only kept the error responses, not the full output, sorry.  I'll take another look at this next week though, head down in something else I need to complete at the moment.   Hopefully have something more by the time you're back from vacation. (have a great time btw :) )

**Comment by stuartleeks** [comment]
> head down in something else I need to complete at the momen
Sure thing

I rebased on latest `master` and addressed the PR feedback and re-running the tests I see same errors you described. If I get chance to look this afternoon I will, otherwise I'll add it to my list for when I'm back from vacation :-)

**Comment by stuartleeks** [comment]
I turned on logging and found the bug. Unfortunately it's in the `giovanni` SDK code I wrote, so once [this PR](https://github.com/tombuildsstuff/giovanni/pull/31) is merged we'll need a release of the SDK and a PR to update the version in the provider

**Comment by jackofallops** [comment]
So... I merged and released the Giovanni fix, and found that there's another issue. The `DeleteResponder` is expecting a 202 for success, but the service returns a 200.  I'll need to loop back around to sort that out when I have a little more time, but I think that's all that's left to progress this.

**Comment by stuartleeks** [comment]
@jackofallops - I'm back from vacation now and may get a little bit of time to look at it this week. Before I get started, I just wanted to check whether you are already looking at this?

**Comment by stuartleeks** [comment]
Have created https://github.com/tombuildsstuff/giovanni/pull/33 to address the path deletion issue.
Also rebased this PR on latest master

**Comment by jackofallops** [comment]
Hi @stuartleeks 
I've been on leave myself on and off, so I've not done anything further (as you likely noticed!) @tombuildsstuff released a new version of Giovanni with your changes in. I think we have some other dependency changes happening in the provider so I'll check back in here when I know for sure so I don't overcomplicate things for anyone.

**Comment by stuartleeks** [comment]
Thanks @jackofallops. I had a quick look and it doesn't look like the updated giovanni is in this repo yet so I was planning to create a PR to do that (which can be merged whenever fits with other changes). Then this PR can be rebased once that's in.

Does that sound like a reasonable approach?

**Comment by jackofallops** [comment]
> Thanks @jackofallops. I had a quick look and it doesn't look like the updated giovanni is in this repo yet so I was planning to create a PR to do that (which can be merged whenever fits with other changes). Then this PR can be rebased once that's in.
> 
> Does that sound like a reasonable approach?

I think Tom's working on the dependency items at the moment, he may well have Giovanni on that list. I'll check with him asap.

**Comment by tombuildsstuff** [comment]
@stuartleeks I've got a WIP PR in-flight for the dependency upgrades - would you mind waiting for that, since the other items need to be updated/tested at the same time?

**Comment by stuartleeks** [comment]
That's even better ;-)

Lmk when that's merged and I'll rebase and re-test this PR locally

**Comment by tombuildsstuff** [comment]
@stuartleeks unfortunately that's blocked on a breaking change in the Networking API's - but I've got a separate thread running that down - will drop a message here when that's fixed however

**Comment by stuartleeks** [comment]
Ah, ok. Is it possible to update the Giovanni version as a standalone change to unblock this while waiting on the Networking API change?

**Comment by tombuildsstuff** [comment]
@stuartleeks not so easily unfortunately - the latest version of tombuildsstuff/giovanni (and hashicorp/go-azure-helpers etc) remove the requirement/redirection of the custom version of Azure/go-autorest - so these all need to be upgraded together

**Comment by stuartleeks** [comment]
Hi @tombuildsstuff - just checking back to see if there's any update on this? Our next engagement is starting to shape up, and as luck would have it, it seems like this capability will be useful in that 😄

**Comment by stuartleeks** [comment]
Hi @tombuildsstuff  - just wanted to check back how this is progressing? Thanks!

**Comment by jackofallops** [comment]
Hi @stuartleeks - Afraid this is blocked still due to the Giovanni upgrade requiring the latest go-autorest, which is blocked on the Azure SDK. We're working with Microsoft to unblock it, but we don't have an ETA yet.

**Comment by mozts2005** [comment]
Is there an ETA on this PR and if help is needed how can I help?

**Comment by stuartleeks** [comment]
@tombuildsstuff  @jackofallops - with [this PR](https://github.com/Azure/go-autorest/pull/567) merged, are we now set to update Giovanni in the provider so that this PR can be merged?

**Comment by tombuildsstuff** [comment]
@stuartleeks yes, AFAIK we should be able to update Giovanni now that go-autorest is fixed 👍

**Comment by stuartleeks** [comment]
Thanks @tombuildsstuff. We have a customer engagement where this will be very useful soon - let me know if there's anything I can help with on Giovanni or updating the version of Giovanni in the provider :-)

**Comment by tombuildsstuff** [comment]
Dependency for this is https://github.com/terraform-providers/terraform-provider-azurerm/pull/9192 - which should be merged shortly/today

**Comment by mozts2005** [comment]
Now that #9192 has been merged. Do we have an ETA?

**Comment by jjcollinge** [comment]
Would be awesome to get this in soon, I also have a customer engagement that will require this.

**Review by tombuildsstuff — APPROVED** [review]
LGTM - thanks for pushing those changes / sorry for the delayed re-review here @stuartleeks

**Comment by tombuildsstuff** [comment]
@stuartleeks hope you don't mind but I've rebased this and pushed a commit to fix the build failure now the shim layer's been merged - I'll kick off the tests but this should otherwise be good to merge 👍

**Comment by stuartleeks** [comment]
Thanks for the rebase @tombuildsstuff! Looks like the tests have all passed :-)

**Comment by katbyte** [comment]
@stuartleeks - it seems the tests for us are failing with:
```
------- Stdout: -------
=== RUN   TestAccAzureRMStorageDataLakeGen2Path_withOwner
=== PAUSE TestAccAzureRMStorageDataLakeGen2Path_withOwner
=== CONT  TestAccAzureRMStorageDataLakeGen2Path_withOwner
    TestAccAzureRMStorageDataLakeGen2Path_withOwner: testing.go:684: Step 0 error: errors during apply:
        
        Error: Error setting access control for Path "testpath" in File System "fstest" in Storage Account "acctestaccu2vbs": datalakestore.Client#SetAccessControl: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationPermissionMismatch" Message="This request is not authorized to perform this operation using this permission.\nRequestId:275ae0c4-a01f-0018-72c5-bd9a16000000\nTime:2020-11-18T16:12:51.1978201Z"
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test372406367/main.tf line 50:
          (source code not available)
        
        
--- FAIL: TestAccAzureRMStorageDataLakeGen2Path_withOwner (143.75s)
FAIL
```

any idea why that could be?

**Comment by stuartleeks** [comment]
@katbyte - ah. Is it possible to assign the account running the tests the `Storage Blob Data Owner` role?

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2020-11-19 at 13 45 57

@stuartleeks as a heads up we ended up pushing a role assignment within the tests, rather than at the subscription level - to be able to differentiate between users who have Storage RP permissions and don't when the shim layer we've added recently is used (to toggle between Data Plane and Resource Manager resources)

**Comment by stuartleeks** [comment]
@tombuildsstuff - nice, I like the approach!

Thanks for getting this merged :-D

**Pull request merged**
